wren-ai-service/src/pipelines/generation/chart_generation.py

```diff
@@ -0,0 +1,278 @@
+
+            return {
+                "results": {
+                    "chart_schema": {},
+                    "reasoning": "",
+                }
+            }
+        except Exception as e:
+            logger.exception(f"JSON deserialization failed: {e}")
+
+            return {
+                "results": {
+                    "chart_schema": {},
+                    "reasoning": "",
+                }
+            }
+
+
+## Start of Pipeline
+@timer
+@observe(capture_input=False)
+def preprocess_data(
+    data: Dict[str, Any], chart_data_preprocessor: ChartDataPreprocessor
+) -> dict:
+    return chart_data_preprocessor.run(data)
+
+
+@timer
+@observe(capture_input=False)
+def prompt(
+    query: str,
+    sql: str,
+    preprocess_data: dict,
+    language: str,
+    prompt_builder: PromptBuilder,
+) -> dict:
+    sample_data = preprocess_data["results"]["sample_data"]
+    sample_data_statistics = preprocess_data["results"]["sample_data_statistics"]
+
+    return prompt_builder.run(
+        query=query,
+        sql=sql,
+        sample_data=sample_data,
+        sample_data_statistics=sample_data_statistics,
+        language=language,
+    )
+
+
+@async_timer
+@observe(as_type="generation", capture_input=False)
+async def generate_chart(prompt: dict, generator: Any) -> dict:
+    return await generator.run(prompt=prompt.get("prompt"))
+
+
+@timer
+@observe(capture_input=False)
+def post_process(
+    generate_chart: dict,
+    vega_schema: Dict[str, Any],
+    post_processor: ChartGenerationPostProcessor,
+) -> dict:
+    return post_processor.run(generate_chart.get("replies"), vega_schema)
+
+
+## End of Pipeline
+class ChartGenerationResults(BaseModel):
+    reasoning: str
+    chart_schema: (
+        LineChartSchema
+        | BarChartSchema
+        | PieChartSchema
+        | GroupedBarChartSchema
+        | StackedBarChartSchema
+        | AreaChartSchema
+    )
+
+
+CHART_GENERATION_MODEL_KWARGS = {
+    "response_format": {
+        "type": "json_schema",
+        "json_schema": {
+            "name": "matched_schema",
+            "schema": ChartGenerationResults.model_json_schema(),
+        },
+    }
+}
+
+
+class ChartGeneration(BasicPipeline):
+    def __init__(
+        self,
+        llm_provider: LLMProvider,
+        **kwargs,
+    ):
+        self._components = {
+            "prompt_builder": PromptBuilder(
+                template=chart_generation_user_prompt_template
+            ),
+            "generator": llm_provider.get_generator(
+                system_prompt=chart_generation_system_prompt,
+                generation_kwargs=CHART_GENERATION_MODEL_KWARGS,
+            ),
+            "chart_data_preprocessor": ChartDataPreprocessor(),
+            "post_processor": ChartGenerationPostProcessor(),
+        }
+        self._configs = {
+            "vega_schema": requests.get(
+                "https://vega.github.io/schema/vega-lite/v5.json"
+            ).json(),
```

Comment:
I have two questions here.
1. should we handle the offline issue what if this web down, it will cause our service can start up
2. how about move the URL to setting object, then we can easily to change the URL for upgrading the version